File: solutions/0303. Range Sum Query - Immutable/0303.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class NumArray {
 public:
  NumArray(const vector<int>& nums) : prefix(nums.size() + 1) {
    partial_sum(nums.begin(), nums.end(), prefix.begin() + 1);
  }
    
  int sumRange(const int left, const int right) {
    return prefix[right + 1] - prefix[left];
  }
  
 private:
  vector<int> prefix;    
};

File: solutions/0303. Range Sum Query - Immutable/0303.py
# Time complexity: O(n)
# Space complexity: O(n)
class NumArray:
  def __init__(self, nums: List[int]):
    self.prefix = [0] + list(itertools.accumulate(nums))

  def sumRange(self, left: int, right: int) -> int:
    return self.refix[right + 1] - self.prefix[left]